### ISSUE TITLE
Provide a way to execute top-level functions before `main` function

### ISSUE THREAD

**Issue by vanesyan** [dart-lang/sdk#34409]
Currently I'm working on a project is aimed to provide a common interface for SQL drivers build for Dart something like JDBC in Java or `database/sql` in Go. I faced a problem where drivers cannot register themselves, so user have to run registration in the `main`  function manually:

```dart
import "package:sql/sql.dart" as sql;
import "package:mysql/mysql.dart" as mysql;

void main() {
  sql.register("mysql", mysql.driver);
}
```

I know that it's possible to use `dart:mirrors` to dynamically invoke function before the `main` execution,
and the second choice is to use codegen that somehow generates those code.
The first solution is out of scope, because of the 2 facts there are mixed signals about `dart:mirrors` stability on VM and Flutter lacks support of mirror system. The second solution is also out of scope for this project because of the facts that it gives end-user more headaches rather than manual registration.

For instance Go solves this problem by providing `init` function that is executed right before `main` function. Particular Go project can have a lot of those `init` functions and all of them are executed in the order of corresponding packages imports (I guess, don't 100% sure about the order of execution).

For concrete example here is part of code from mysql driver for Go:
https://github.com/go-sql-driver/mysql/blob/7ac0064e822156a17a6b598957ddf5e0287f8288/driver.go#L163-L165

For the second example for solution of this problem, in C++ we can use well-known hack - instantiation of class with custom constructor assigned to top-level variable, e.g.:
```c++
#include <iostream>

class A {
public:
    A() {
        std::cout << "Before `main`" << std::endl;
    }
};

A a;

int main() {
    std::cout << "`main`" << std::endl;
}
```

**Comment by zoechi** [comment]
Why does the driver need to be initialized before main?
If you have a MySql class you could add the initialization code to the constructor.
What is the problem with requiring `init` to be called?
What if only parts of the app use SQL access and the user prefers the driver to be only initialized on first use?

**Comment by ajrcarey** [comment]
Another example would be immutable deserialization. The only real way to handle this is to have a registry of deserialization handlers available before the first deserialization operation begins. Sure, you could ask the user to handle registration of these serializers manually in main() before they deserialize anything, or you could try and have codegen do it at build time, but it's more convenient for everyone if deserializers can self-register using static initialization before main() is executed.

**Comment by vanesyan** [comment]
This code prints empty list (`[]`) for me:

```dart
final a = <int>[];

class B {
    B() {
        a.add(1);
    }
}

final b = new B();

void main() {
    print(a);
}
```

The main idea of register driver automatically is the conveniency of use. The end-user always would like to worry a little as possible about such bootstrap parts of code. The second one idea of `init` function is possible extra bootstrap steps of driver package before the registration, the end-user really shouldn't worry about.

**Comment by lrhn** [comment]
There is no way in Dart to have code executed without it being called directly or indirectly from the `main` function of the entry library of the isolate.

It is a deliberate design choice for Dart that there is no initialization step before starting `main`. It avoids an uncontrollable amount of computation happening outside of any context, and when Dart is run in a browser, it prevents arbitrary initialization from delaying page loading.

That is why static variables are lazily initialized. It allows you to initialize a variable before it's used, but it isn't initialized until then.

There is no way for some code somewhere to have an effect if the code is not run explicitly.

You say:

> I know that it's possible to use dart:mirrors to dynamically invoke function before the main execution,

I'd like to know how that is done, because it shouldn't be possible.

**Comment by kika** [comment]
@lrhn For the record, this: 

> It is a deliberate design choice for Dart that there is no initialization step before starting main. It avoids an uncontrollable amount of computation happening outside of any context, and when Dart is run in a browser, it prevents arbitrary initialization from delaying page loading.

appears to be no longer true. Flutter runs `registerPlugins()` before the main. Surprisingly it's the same functionality that the OP was asking for :-D
